util/perftests.h

```diff
@@ -0,0 +1,73 @@
+#ifndef _PERFTESTS_H_
+#define _PERFTESTS_H_
+
+#include <stddef.h>
+
+/**
+ * PERFTESTS_*:
+ * These macros handle various "housekeeping" functions.  To time a piece of
+ * code and print a message to stdout, use:
+ *
+ *     PERFTESTS_VAR
+ *
+ *     ... general initialization...
+ *
+ *     PERFTESTS_LOOP_OUTER_BEGIN
+ *             ... per-buffer-size initialization...
+ *
+ *             PERFTESTS_LOOP_INNER_BEGIN
+ *                     ... code you want to time...
+ *             PERFTESTS_LOOP_INNER_END
+ *
+ *     PERFTESTS_LOOP_OUTER_END
+ *
+ * These macros should not have a colon ; appended to them.
+ */
+
+/* These variables are prepended with "perftests_". */
+#define PERFTESTS_VARS(err_label_num)					\
+	uint8_t * perftests_buf;					\
+	struct timeval perftests_begin, perftests_end;			\
+	double perftests_delta_s;					\
+	size_t perftests_i, perftests_j;				\
+	size_t perftests_bufsize;					\
+	size_t perftests_num_buffers;					\
+	const size_t perftests_num =					\
+	     sizeof(PERFTESTS_SIZES) / sizeof(PERFTESTS_SIZES[0]);	\
+									\
+	if ((perftests_buf = malloc(PERFTESTS_MAX_BUFSIZE)) == NULL) {	\
+		warnp("malloc");					\
+		goto err ## err_label_num;				\
+	}								\
+	memset(perftests_buf, 0, PERFTESTS_MAX_BUFSIZE);
+
+#define PERFTESTS_LOOP_OUTER_BEGIN					\
+	for (perftests_i = 0; perftests_i < perftests_num; perftests_i++) { \
+		perftests_bufsize = PERFTESTS_SIZES[perftests_i];	\
+		perftests_num_buffers = PERFTESTS_BYTES / perftests_bufsize; \
+		memset(perftests_buf, 0, perftests_bufsize);
+
+#define PERFTESTS_LOOP_INNER_BEGIN(err_label_num)			\
+	if (monoclock_get_cputime(&perftests_begin)) {			\
+		warnp("monoclock_get_cputime()");			\
+		goto err ## err_label_num;				\
+	}								\
+	for (perftests_j = 0; perftests_j < perftests_num_buffers;	\
+	    perftests_j++) {
+
+#define PERFTESTS_LOOP_INNER_END(err_label_num)				\
+	}								\
+	if (monoclock_get_cputime(&perftests_end)) {			\
+		warnp("monoclock_get_cputime()");			\
+		goto err ## err_label_num;				\
+	}								\
+	perftests_delta_s = timeval_diff(perftests_begin, perftests_end); \
+	printf("... in %zu blocks of size %zu:\t%.02f s\t%.01f MB/s\n",	\
+	    perftests_num_buffers, perftests_bufsize, perftests_delta_s,\
+	    (double)PERFTESTS_BYTES / 1e6 / perftests_delta_s);		\
+	fflush(stdout);
+
+#define PERFTESTS_LOOP_OUTER_END					\
+	}
```

Comment:
This also needs to `free(perftests_buf)`.

(I was experimenting with `uint8_t perftests_buf[PERFTESTS_MAX_BUFSIZE];` but changed my mind about that a few minutes before uploading the PR, and I forgot to fix that until I saw it on github.)